### PR TITLE
Replace stream with cycles

### PR DESCRIPTION
**Pull request by kostya05983**
Work in progress, change streams to for loops. Please see [https://github.com/spring-projects/spring-security/issues/7154](url)
gh-7154

### CODE REVIEW

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java [hunk 1]
```diff
@@ -64,10 +64,17 @@ public DelegatingOAuth2AuthorizedClientProvider(List<OAuth2AuthorizedClientProvi
 	@Nullable
 	public OAuth2AuthorizedClient authorize(OAuth2AuthorizationContext context) {
 		Assert.notNull(context, "context cannot be null");
-		return this.authorizedClientProviders.stream()
-				.map(authorizedClientProvider -> authorizedClientProvider.authorize(context))
-				.filter(Objects::nonNull)
-				.findFirst()
-				.orElse(null);
+		final List<OAuth2AuthorizedClient> clients = new ArrayList<>();
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java]: No need for a list. Just return the first non null value

Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java]: fix this.

**Comment by kostya05983** [comment]
Hello, @rwinch I think, it's ready for review. Now, you may review it.

**Comment by kostya05983** [comment]
I updated changes from master.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java [hunk 2]
```diff
@@ -401,5 +398,11 @@ private String buildAuthorizationRequestUri() {
 					.build()
 					.toUriString();
 		}
+
+		private LinkedHashSet<String> toLinkedHashSet(String... scope) {
+			final LinkedHashSet<String> result = new LinkedHashSet<>();
```
Comment (jzheaux) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java]: To conform with other Spring Security code, please remove the `final` keyword from in front of local variables.

File: config/src/main/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfiguration.java [hunk 3]
```diff
@@ -175,6 +171,20 @@ public void setObjectPostProcessor(ObjectPostProcessor<Object> objectPostProcess
 		return (T) proxyFactory.getObject();
 	}
 
+	private List<String> getPrimaryBeanNames(String[] beanNamesForType) {
+		final List<String> list = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply config/src/main/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfiguration.java]: Fix it

File: core/src/main/java/org/springframework/security/converter/RsaKeyConverters.java [hunk 4]
```diff
@@ -16,17 +16,14 @@
 
 package org.springframework.security.converter;
 
-import java.io.BufferedReader;
-import java.io.InputStream;
-import java.io.InputStreamReader;
+import java.io.*;
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply core/src/main/java/org/springframework/security/converter/RsaKeyConverters.java]: thanks, expand all

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java [hunk 5]
```diff
@@ -64,10 +63,17 @@ public DelegatingOAuth2AuthorizedClientProvider(List<OAuth2AuthorizedClientProvi
...
+		for (OAuth2AuthorizedClientProvider authorizedClientProvider: authorizedClientProviders) {
+			final OAuth2AuthorizedClient auth2AuthorizedClient = authorizedClientProvider.authorize(context);
+			if (auth2AuthorizedClient != null) {
+				clients.add(auth2AuthorizedClient);
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java]: Thanks, changed it

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java [hunk 6]
```diff
@@ -64,10 +63,17 @@ public DelegatingOAuth2AuthorizedClientProvider(List<OAuth2AuthorizedClientProvi
...
 	public OAuth2AuthorizedClient authorize(OAuth2AuthorizationContext context) {
 		Assert.notNull(context, "context cannot be null");
-		return this.authorizedClientProviders.stream()
-				.map(authorizedClientProvider -> authorizedClientProvider.authorize(context))
-				.filter(Objects::nonNull)
-				.findFirst()
-				.orElse(null);
+		final List<OAuth2AuthorizedClient> clients = new ArrayList<>();
+		for (OAuth2AuthorizedClientProvider authorizedClientProvider: authorizedClientProviders) {
+			final OAuth2AuthorizedClient auth2AuthorizedClient = authorizedClientProvider.authorize(context);
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java]: ok

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java [hunk 7]
```diff
@@ -127,10 +128,31 @@ public String toString() {
...
+		sb.append(list.get(list.size() - 1).getAuthority());
+		return sb.toString();
+	}
+
+	private String collectAttributes(Set<Map.Entry<String, Object>> entrySet) {
+		final List<Map.Entry<String, Object>> list = new ArrayList<>(entrySet);
+		final StringBuilder sb = new StringBuilder();
+
+		for (int i = 0; i< list.size() - 1; i++) {
+			sb.append(list.get(i).getKey()).append("=").append(list.get(i).getValue()).append(", ");
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java]: fix

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/BearerTokenAuthenticationEntryPoint.java [hunk 8]
```diff
@@ -112,13 +111,20 @@ public final void setRealmName(String realmName) {
...
+		wwwAuthenticate.append("Bearer");
 		if (!parameters.isEmpty()) {
-			wwwAuthenticate += parameters.entrySet().stream()
-					.map(attribute -> attribute.getKey() + "=\"" + attribute.getValue() + "\"")
-					.collect(Collectors.joining(", ", " ", ""));
+			wwwAuthenticate.append(" ");
+			int i = 0;
+			for (Map.Entry<String, String> entry : parameters.entrySet()) {
+				wwwAuthenticate.append(entry.getKey()).append("=\"").append(entry.getValue()).append("\"");
+						if (i != parameters.size() - 1) {
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/BearerTokenAuthenticationEntryPoint.java]: No, there isn't a reason, There are three cases where we use such function and I change it in one place and copy it in another and my tabs was broken

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java (hunk 2 above)
Comment (kostya05983) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java]: Removed.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jzheaux) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java]: It looks like this feedback has not yet been addressed. My I suggest:

```suggestion
			OAuth2AuthorizedClient outh2AuthorizedClient = authorizedClientProvider.authorize(context);
```

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java (hunk 1 above)
Comment (jzheaux) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java]: Let's remove the list completely here. Remove this line and change the for loop to return when it finds the first value.

Something like:

```java
for (OAuth2AuthorizedClientProvider authorizedClientProvider: authorizedClientProviders) {
    OAuth2AuthorizedClient oauth2AuthorizedClient = authorizedClientProvider.authorize(context);
    if (oauth2AuthorizedClient != null) {
        return oauth2AuthorizedClient;
    }
}
return null;
```

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/converter/ObjectToListStringConverter.java [hunk 9]
```diff
@@ -64,10 +63,13 @@ public Object convert(Object source, TypeDescriptor sourceType, TypeDescriptor t
 			}
 		}
 		if (source instanceof Collection) {
-			return ((Collection<?>) source).stream()
-					.filter(Objects::nonNull)
-					.map(Objects::toString)
-					.collect(Collectors.toList());
+			final Collection<String> results = new ArrayList<>();
```
Comment (jzheaux) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/converter/ObjectToListStringConverter.java]: For consistency across the project, please remove the `final` keyword from this local variable.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java [hunk 10]
```diff
@@ -127,10 +128,34 @@ public String toString() {
 		sb.append("Name: [");
 		sb.append(this.getName());
 		sb.append("], Granted Authorities: [");
-		sb.append(this.getAuthorities().stream().map(GrantedAuthority::getAuthority).collect(Collectors.joining(", ")));
```
Comment (jzheaux) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java]: I think this might be simpler as:

```java
sb.append("], GrantedAuthorities: ");
sb.append(getAuthorities())
sb.append(", User Attributes: ");
sb.append(getAttributes());
return sb.toString();
```

File: web/src/main/java/org/springframework/security/web/server/header/CompositeServerHttpHeadersWriter.java [hunk 11]
```diff
@@ -43,8 +42,10 @@ public CompositeServerHttpHeadersWriter(List<ServerHttpHeadersWriter> writers) {
 
 	@Override
 	public Mono<Void> writeHttpHeaders(ServerWebExchange exchange) {
-		Stream<Mono<Void>> results = writers.stream().map( writer -> writer.writeHttpHeaders(exchange));
-		return Mono.when(results.collect(Collectors.toList()));
+		List<Mono<Void>> results = new ArrayList<>();
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/server/header/CompositeServerHttpHeadersWriter.java]: I appreciate that you used an interface as your reference type here (e.g. `List` instead of `ArrayList`).

File: web/src/main/java/org/springframework/security/web/server/header/ClearSiteDataServerHttpHeadersWriter.java [hunk 12]
```diff
@@ -81,9 +78,12 @@ public String getHeaderValue() {
 	}
 
 	private String transformToHeaderValue(Directive... directives) {
-		return Stream.of(directives)
-				.map(Directive::getHeaderValue)
-				.collect(Collectors.joining(", "));
+		final StringBuilder sb = new StringBuilder();
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/server/header/ClearSiteDataServerHttpHeadersWriter.java]: For consistency across the project, please remove the `final` keyword here.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java [hunk 13]
```diff
@@ -64,10 +63,12 @@ public DelegatingOAuth2AuthorizedClientProvider(List<OAuth2AuthorizedClientProvi
 	@Nullable
 	public OAuth2AuthorizedClient authorize(OAuth2AuthorizationContext context) {
 		Assert.notNull(context, "context cannot be null");
-		return this.authorizedClientProviders.stream()
-				.map(authorizedClientProvider -> authorizedClientProvider.authorize(context))
-				.filter(Objects::nonNull)
-				.findFirst()
-				.orElse(null);
+		for (OAuth2AuthorizedClientProvider authorizedClientProvider: authorizedClientProviders) {
+			final OAuth2AuthorizedClient oauth2AuthorizedClient = authorizedClientProvider.authorize(context);
```
Comment (jzheaux) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java]: Thanks for simplifying this. Will you please also remove the `final` keyword from the local variable here?

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java [hunk 14]
```diff
@@ -137,9 +136,15 @@ public void setClockSkew(Duration clockSkew) {
 	}
 
 	private static OAuth2Error invalidIdToken(Map<String, Object> invalidClaims) {
-		String claimsDetail = invalidClaims.entrySet().stream()
-				.map(it -> it.getKey() + " (" + it.getValue() + ")")
-				.collect(Collectors.joining(", "));
+		StringBuilder claimsDetail = new StringBuilder();
```
Comment (jzheaux) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java]: Actually, since this is simply going to an error message, let's simply rely on the `toString` method. The code will be easier to read and to maintain:

```java
return new OAuth2Error("invalid_id_token",
    "The ID Token contains invalid claims: " + invalidClaims,
// ...
```

Also, your unit test will also need an update.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryReactiveClientRegistrationRepository.java [hunk 15]
```diff
@@ -79,4 +76,12 @@ public InMemoryReactiveClientRegistrationRepository(List<ClientRegistration> reg
 	public Iterator<ClientRegistration> iterator() {
 		return this.clientIdToClientRegistration.values().iterator();
 	}
+
+	private ConcurrentHashMap<String, ClientRegistration> toConcurrentMap(List<ClientRegistration> registrations) {
+		final ConcurrentHashMap<String, ClientRegistration> result = new ConcurrentHashMap<>();
+		for (ClientRegistration registration : registrations) {
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryReactiveClientRegistrationRepository.java]: Change it

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java (hunk 10 above)
Comment (kostya05983) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java]: I can't change function names from collectAuthorities to getAuthorities, cause there are exists overriding functions with this names, but I change the params.

File: web/src/main/java/org/springframework/security/web/server/header/CompositeServerHttpHeadersWriter.java (hunk 11 above)
Comment (kostya05983) [reply web/src/main/java/org/springframework/security/web/server/header/CompositeServerHttpHeadersWriter.java]: Thanks

File: core/src/main/java/org/springframework/security/core/userdetails/MapReactiveUserDetailsService.java [hunk 16]
```diff
@@ -56,7 +55,11 @@ public MapReactiveUserDetailsService(UserDetails... users) {
 	 */
 	public MapReactiveUserDetailsService(Collection<UserDetails> users) {
 		Assert.notEmpty(users, "users cannot be null or empty");
-		this.users = users.stream().collect(Collectors.toConcurrentMap( u -> getKey(u.getUsername()), Function.identity()));
+		final ConcurrentHashMap<String, UserDetails> map = new ConcurrentHashMap<>();
+		for (UserDetails user: users) {
+			map.put(user.getUsername(), user);
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply core/src/main/java/org/springframework/security/core/userdetails/MapReactiveUserDetailsService.java]: I added one

File: core/src/main/java/org/springframework/security/core/userdetails/MapReactiveUserDetailsService.java [hunk 17]
```diff
@@ -56,7 +55,11 @@ public MapReactiveUserDetailsService(UserDetails... users) {
 	 */
 	public MapReactiveUserDetailsService(Collection<UserDetails> users) {
 		Assert.notEmpty(users, "users cannot be null or empty");
-		this.users = users.stream().collect(Collectors.toConcurrentMap( u -> getKey(u.getUsername()), Function.identity()));
+		final ConcurrentHashMap<String, UserDetails> map = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply core/src/main/java/org/springframework/security/core/userdetails/MapReactiveUserDetailsService.java]: ok

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryClientRegistrationRepository.java [hunk 18]
```diff
@@ -62,9 +57,19 @@ public InMemoryClientRegistrationRepository(List<ClientRegistration> registratio
...
 	private static Map<String, ClientRegistration> createRegistrationsMap(List<ClientRegistration> registrations) {
 		Assert.notEmpty(registrations, "registrations cannot be empty");
-		Collector<ClientRegistration, ?, ConcurrentMap<String, ClientRegistration>> collector =
-				toConcurrentMap(ClientRegistration::getRegistrationId, Function.identity());
-		return registrations.stream().collect(collectingAndThen(collector, Collections::unmodifiableMap));
+		return toUnmodifiableConcurrentMap(registrations);
+	}
+
+	private static Map<String, ClientRegistration> toUnmodifiableConcurrentMap(List<ClientRegistration> registrations) {
+		final ConcurrentHashMap<String, ClientRegistration> result = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryClientRegistrationRepository.java]: remove

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryReactiveClientRegistrationRepository.java [hunk 19]
```diff
@@ -79,4 +76,12 @@ public InMemoryReactiveClientRegistrationRepository(List<ClientRegistration> reg
 	public Iterator<ClientRegistration> iterator() {
 		return this.clientIdToClientRegistration.values().iterator();
 	}
+
+	private ConcurrentHashMap<String, ClientRegistration> toConcurrentMap(List<ClientRegistration> registrations) {
+		final ConcurrentHashMap<String, ClientRegistration> result = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryReactiveClientRegistrationRepository.java]: remove

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java (hunk 14 above)
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java]: change and update a unit test

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java (hunk 13 above)
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/DelegatingOAuth2AuthorizedClientProvider.java]: remove

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryClientRegistrationRepository.java [hunk 20]
```diff
@@ -62,9 +57,19 @@ public InMemoryClientRegistrationRepository(List<ClientRegistration> registratio
...
+		for (ClientRegistration registration : registrations) {
+			if (result.containsKey(registration.getRegistrationId())) {
```
[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryClientRegistrationRepository.java]: Oks, also I need a remove test for it, because there is a test which checks duplicates.

[the remark this answers is not in the document]
Comment (kostya05983) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryClientRegistrationRepository.java]: Removed

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/converter/ObjectToListStringConverter.java (hunk 9 above)
Comment (kostya05983) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/converter/ObjectToListStringConverter.java]: Removed

**Comment by kostya05983** [comment]
@jzheaux Thanks for guidance. I think it's ready, but I tried to rebase and squash my nearest commits into one,in such groups: [Replace stream, First version of replacing streams, fix wwAuthentication and code style] and other similar group with other dates. But I got a big diff on test branch.  Can you tell me how I can squash into one my commits which between others commits and not to danger history. I tried with interactive git rebase.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryClientRegistrationRepository.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (jzheaux) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/InMemoryClientRegistrationRepository.java]: @kostya05983 for some reason I'd missed that there is a test that relies on this functionality. I apologize, but because of that, we should add the duplicate checks as well as the test back in. My mistake.

**Comment by jzheaux** [comment]
To rebase, you can do something like (assuming that you've called Spring Security's upstream `spring-projects`)

```
git checkout master
git pull spring-projects master
git checkout gh-7154-replace-streams
git rebase master
```

You will likely have some conflicts to resolve, but at that point, the commits on your feature branch should be at the end of the chain. Once you've resolved the conflicts, then do (assuming your fork is called `origin`):

```
git push origin +gh-7154-replace-streams
```

**Comment by kostya05983** [comment]
@jzheaux , I squashed commits, it's ready for merge

**Comment by jzheaux** [comment]
@kostya05983 - thank you for your careful work on this. This is now merged into `master` via f6c650db47233cadf8a72f6bc3ac667c1b50e92f.

I edited the commit message in order to align with the [contributor guidelines](https://github.com/spring-projects/spring-security/blob/master/CONTRIBUTING.md#format-commit-messages).
